datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -332,12 +336,19 @@ private synchronized void startEventThread() {
     }
   }
 
-  private synchronized boolean stopEventThread() {
+  private boolean stopEventThread() {
     // interrupt the thread if it's not gracefully shutdown
     while (_eventThread.isAlive()) {
```

Comment:
Is it ok to not synchronize the thread state check here ? What is it that we are actually synchronizing in this method ?

Reply:
I think the synchronized block should be around while loop with cv. Rest sounds good.

Reply:
do you mean something like this ??

```  
private boolean stopEventThread() {
  synchronized (_conditionalVariableForCoordinatorObjectSynchronization) {
    _conditionalVariableForCoordinatorObjectSynchronization.wait(duration.toMillis());
    while (_eventThread.isAlive()) {
      try {
        synchronized (this) {
          _eventThread.interrupt();
          _eventThread.join(EVENT_THREAD_SHORT_JOIN_TIMEOUT);
        }
      } catch (InterruptedException e) {
        return true;
      }
    }
  }
  return false;
}
  ```

Reply:
If we define a lock explicitly (comment above), we can only use one lock for cv and also the object. You will have to replace all synchronize constructs on a method/object with this lock. That would be much cleaner, IMO ? Wdyt ?

Reply:
> I think the synchronized block should be around while loop with cv. 
We cannot have the conditional variable's synchronized block around the while loop. The conditional variable block has to release the lock after calling the wait method so that the notify method can acquire the lock on the conditional variable block.

Reply:
> If we define a lock explicitly (comment above), we can only use one lock for cv and also the object. You will have to replace all synchronize constructs on a method/object with this lock. That would be much cleaner, IMO ? Wdyt ?

We can define another lock explicitly I guess (instead of sync blocks), but we cannot combine both the locks (conditional variable lock and the coordinator object's lock). 

With the conditional variable we are making sure of the following: 
1. The Coordinator Event thread never gets halted on conditional variable.
2. The "ZK Session Expiry callback thread" or "the main server thread" (trying to stop the coordinator) gets halted on the conditional variable. It waits for about the heartbeat period for "the coordinator event thread" to finish the event handling and then "the coordinator event thread" itself would notify these waiting threads to acquire the coordinator object's lock, before "the coordinator event thread" enter's handleEvent again and acquires the lock again. 

I have added two different test scenarios with detailed comments, it might explain it more clearly.

Reply:
I looked up and there was an intrinsic conditional variable of the class which we could use along with keeping the method definition synchronized. While waiting on that intrinsic CV, the lock will be released and that is how we can let the other thread to be prioritized. 

With this change, it made the change very cleaner. Thank you for the insights @atoomula!